File: Digital Business Enablement/buy.it-api/src/main/java/br/com/fiap/buy/it/service/AvaliacaoService.java
package br.com.fiap.buy.it.service;

import br.com.fiap.buy.it.dto.AvaliacaoDTO;
import br.com.fiap.buy.it.model.Avaliacao;
import br.com.fiap.buy.it.repository.AvaliacaoRepository;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.Objects;

@Service
public class AvaliacaoService {

    @Autowired
    private AvaliacaoRepository avaliacaoRepository;

    @Autowired
    private CotacaoService cotacaoService;

    public Page<AvaliacaoDTO> listAll(Pageable pageRequest) {
        Page<Avaliacao> list = avaliacaoRepository.findAll(pageRequest);
        return list.map(this::convertToDto);
    }

    public AvaliacaoDTO findById(Long id) {
        Avaliacao entity = findEntityById(id);
        return convertToDto(entity);
    }

    @Transactional
    public AvaliacaoDTO create(AvaliacaoDTO newData) {
        Avaliacao entity = convertToEntity(newData);
        Avaliacao savedEntity = avaliacaoRepository.save(entity);
        return convertToDto(savedEntity);
    }

    @Transactional
    public AvaliacaoDTO update(Long id, AvaliacaoDTO updatedData) {
        findEntityById(id);
        updatedData.setId(id);
        Avaliacao updatedEntity = convertToEntity(updatedData);    
        Avaliacao savedEntity = avaliacaoRepository.save(updatedEntity);
        return convertToDto(savedEntity);
    }
    
    @Transactional
    public void delete(Long id) {
        Avaliacao entity = findEntityById(id);
        avaliacaoRepository.delete(entity);
    }

    public Avaliacao findEntityById(Long id) {
        return avaliacaoRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "(" + getClass().getSimpleName() + ") - Avaliacao não encontrado(a) por ID: " + id));
    }

    public Page<AvaliacaoDTO> findByCotacaoId(Long cotacaoId, Pageable pageable) {
        if (cotacaoId == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - ID da Cotação não pode ser nulo.");
        }
        Page<Avaliacao> list = avaliacaoRepository.findByCotacaoId(cotacaoId, pageable);
        return list.map(this::convertToDto);
    }

    public Page<AvaliacaoDTO> findByUsuarioId(Long usuarioId, Pageable pageable) {
        if (usuarioId == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - ID do Usuário não pode ser nulo.");
        }
        Page<Avaliacao> list = avaliacaoRepository.findByUsuarioId(usuarioId, pageable);
        return list.map(this::convertToDto);
    }
    
    private AvaliacaoDTO convertToDto(Avaliacao entity) {
        AvaliacaoDTO dto = new AvaliacaoDTO();
        dto.setId(entity.getId());
        dto.setIdCotacao(entity.getCotacao() != null ? entity.getCotacao().getId() : null);
        dto.setData(entity.getData());
        dto.setNotaEntrega(entity.getNotaEntrega());
        dto.setNotaQualidade(entity.getNotaQualidade());
        dto.setNotaPreco(entity.getNotaPreco());
        dto.setDescricao(entity.getDescricao());
        return dto;
    }

    private Avaliacao convertToEntity(AvaliacaoDTO dto) {
        if (Objects.isNull(dto)) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - AvaliacaoDTO não pode ser nulo.");
        }
        Avaliacao entity;
        if (dto.getId() != null) {
            entity = findEntityById(dto.getId());
        } else {
            entity = new Avaliacao();
        }
        if (dto.getIdCotacao() == null) {
            throw new IllegalArgumentException("(" + getClass().getSimpleName() + ") - ID Cotacao não pode ser nulo.");
        }
        entity.setCotacao(cotacaoService.findEntityById(dto.getIdCotacao()));
        entity.setData(dto.getData());
        entity.setNotaEntrega(dto.getNotaEntrega());
        entity.setNotaQualidade(dto.getNotaQualidade());
        entity.setNotaPreco(dto.getNotaPreco());
        entity.setDescricao(dto.getDescricao());
        return entity;
    }
}

File: Enterprise Application Development/buyit/Services/AvaliacaoService.cs
using Buyit.Context;
using Buyit.Dtos;
using Buyit.Models;
using Microsoft.EntityFrameworkCore;

namespace Buyit.Services;
public class AvaliacaoService
{
    private readonly BuyitContext _context;

    public AvaliacaoService(BuyitContext context)
    {
        _context = context;
    }

    public async Task<List<AvaliacaoDto>> FindAll()
    {
        var list = await _context.Avaliacao
            .Include(x => x.Cotacao)
            .ToListAsync();

        return list.Select(entity => ConvertToDto(entity)).ToList();
    }

    public async Task<AvaliacaoDto> FindById(long id)
    {
        var entity = await FindEntityById(id);
        return ConvertToDto(entity);
    }

    public async Task<AvaliacaoDto> Create(AvaliacaoDto newData)
    {
        var entity = await ConvertToEntity(newData);
        _context.Avaliacao.Add(entity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task<AvaliacaoDto> Update(long id, AvaliacaoDto updatedData)
    {
        var entity = await FindEntityById(id);
        updatedData.Id = entity.Id;
        var updatedEntity = await ConvertToEntity(updatedData);
        _context.Entry(entity).CurrentValues.SetValues(updatedEntity);
        await _context.SaveChangesAsync();
        return ConvertToDto(entity);
    }

    public async Task Delete(long id)
    {
        var entity = await FindEntityById(id);
        _context.Avaliacao.Remove(entity);
        await _context.SaveChangesAsync();
    }

    public async Task<AvaliacaoModel> FindEntityById(long id)
    {
        var entity = await _context.Avaliacao
            .Include(x => x.Cotacao)
            .FirstOrDefaultAsync(x => x.Id == id);
        if (entity == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {id}");
        }
        return entity;
    }

    public async Task<List<AvaliacaoDto>> FindByCotacaoId(long id)
    {
        var list = await _context.Avaliacao
            .Where(x => x.Cotacao.Id == id)
            .Include(x => x.Cotacao)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    public async Task<List<AvaliacaoDto>> FindByUsuarioId(long id)
    {
        var list = await _context.Avaliacao
            .Where(x => x.Cotacao.Comprador.Id == id)
            .Include(x => x.Cotacao)
            .ToListAsync();

        return list.Select(x => ConvertToDto(x)).ToList();
    }

    private AvaliacaoDto ConvertToDto(AvaliacaoModel entity)
    {
        return new AvaliacaoDto
        {
            Id = entity.Id,
            IdCotacao = entity.Cotacao.Id,
            Data = entity.Data,
            NotaEntrega = entity.NotaEntrega,
            NotaQualidade = entity.NotaQualidade,
            NotaPreco = entity.NotaPreco,
            Descricao = entity.Descricao
        };
    }

    private async Task<AvaliacaoModel> ConvertToEntity(AvaliacaoDto dto)
    {
        var cotacao = await _context.Cotacao.FindAsync(dto.IdCotacao);
        if (cotacao == null)
        {
            throw new KeyNotFoundException($"Objeto não encontrado com o ID: {dto.IdCotacao}");
        }

        return new AvaliacaoModel
        {
            Id = dto.Id ?? 0,
            Cotacao = cotacao,
            Data = dto.Data,
            NotaEntrega = dto.NotaEntrega,
            NotaQualidade = dto.NotaQualidade,
            NotaPreco = dto.NotaPreco,
            Descricao = dto.Descricao
        };
    }
}